tests/distributions/test_gaussian_mixtures.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+
+import torch
+
+import pytest
+from pyro.distributions import MixtureOfDiagNormalsSharedCovariance
+from tests.common import assert_equal
+
+
+@pytest.mark.parametrize('mix_dist', [MixtureOfDiagNormalsSharedCovariance])
+@pytest.mark.parametrize('K', [3])
+@pytest.mark.parametrize('D', [2, 4])
+@pytest.mark.parametrize('batch_mode', [True, False])
+@pytest.mark.parametrize('flat_logits', [True, False])
+@pytest.mark.parametrize('cost_function', ['quadratic'])
+def test_mean_gradient(K, D, flat_logits, cost_function, mix_dist, batch_mode):
+    n_samples = 2000 * 1000
+    if batch_mode:
+        sample_shape = torch.Size(())
+    else:
+        sample_shape = torch.Size((n_samples,))
+    locs = torch.tensor(torch.rand(K, D), requires_grad=True)
+    sigmas = torch.ones(D) + 0.5 * torch.rand(D)
+    sigmas = torch.tensor(sigmas, requires_grad=True)
+    if not flat_logits:
+        logits = torch.tensor(1.5 * torch.rand(K), requires_grad=True)
+    else:
+        logits = torch.tensor(0.1 * torch.rand(K), requires_grad=True)
+    omega = torch.tensor(0.2 * torch.ones(D) + 0.1 * torch.rand(D), requires_grad=False)
+
+    _pis = torch.exp(logits)
+    pis = _pis / _pis.sum()
+
+    if cost_function == 'cosine':
+        analytic1 = torch.cos((omega * locs).sum(-1))
+        analytic2 = torch.exp(-0.5 * torch.pow(omega * sigmas, 2.0).sum(-1))
+        analytic = (pis * analytic1 * analytic2).sum()
+        analytic.backward()
+    elif cost_function == 'quadratic':
+        analytic = torch.pow(sigmas, 2.0).sum(-1) + torch.pow(locs, 2.0).sum(-1)
+        analytic = (pis * analytic).sum()
+        analytic.backward()
+
+    analytic_grads = {}
+    analytic_grads['locs'] = locs.grad.clone()
+    analytic_grads['scale'] = sigmas.grad.clone()
+    analytic_grads['logits'] = logits.grad.clone()
+    assert locs.grad.shape == locs.shape
+    assert sigmas.grad.shape == sigmas.shape
+    assert logits.grad.shape == logits.shape
+
+    sigmas.grad.zero_()
+    logits.grad.zero_()
+    locs.grad.zero_()
+
+    if mix_dist == MixtureOfDiagNormalsSharedCovariance:
+        params = {'locs': locs, 'scale': sigmas, 'logits': logits}
+        if batch_mode:
+            locs = locs.unsqueeze(0).expand(n_samples, K, D)
```

Comment:
I don't think you need these `.unsqeeze(0)`s. If you generalize from `batch_mode` to `sample_shape` you should be able to unconditionally
```py
locs = locs.expand(sample_shape + (K, D))
sigmas = sigmas.expand(sample_shape + (D,))
logits = logits.expand(sample_shape + (K,))
```

Reply:
i don't think i have time to reason carefully about these kinds of proposed changes before i leave

Reply:
Ok, I'll take that to me that you volunteer to maintain this complex test as we refactor and update due to PyTorch changes.

Reply:
yes i volunteer